contracts/AelinUpFrontDeal.sol

```diff
@@ -0,0 +1,357 @@
+                        nftCollectionRules[i].collectionAddress,
+                        nftCollectionRules[i].purchaseAmount,
+                        nftCollectionRules[i].purchaseAmountPerToken,
+                        nftCollectionRules[i].tokenIds,
+                        nftCollectionRules[i].minTokensEligible
+                    );
+                }
+                hasNftList = true;
+            } else {
+                revert("collection is not compatible");
+            }
+        }
+
+        // deposit underlying token logic
+        // check if the underlyingDealAmount is more than 0,
+        // if yes, transfer it to this contract and store it in `currentDealTokenTotal` and add it
+        if (_depositUnderlayingAmount > 0) {
+            require(
+                IERC20(_dealData.underlyingDealToken).balanceOf(dealCreator) >= _depositUnderlayingAmount,
+                "not enough balance"
+            );
+            IERC20(_dealData.underlyingDealToken).transferFrom(dealCreator, address(this), _depositUnderlayingAmount);
+            uint256 currentDealTokenTotal = IERC20(_dealData.underlyingDealToken).balanceOf(address(this));
+            if (currentDealTokenTotal >= _dealData.underlyingDealTokenTotal) {
+                underlyingDepositComplete = true;
+                purchaseExpiry = block.timestamp + dealData.purchaseDuration;
+                emit DealFullyFunded(address(this), block.timestamp, purchaseExpiry);
+            }
+
+            emit DepositDealToken(_dealData.underlyingDealToken, dealCreator, currentDealTokenTotal);
+        }
+    }
+
+    /**
+     * @dev method for holder to deposit underlying deal tokens
+     * all underlying deal tokens must be deposited for the purchasing period to start
+     * if tokens were deposited directly, this method must still be called to start the purchasing period
+     */
+    function depositUnderlyingTokens(uint256 _depositUnderlyingAmount) public onlyHolder {
+        UpFrontDeal memory _dealData = dealData;
+
+        require(
+            IERC20(_dealData.underlyingDealToken).balanceOf(msg.sender) >= _depositUnderlyingAmount,
+            "not enough balance"
+        );
+        require(!underlyingDepositComplete, "already deposited the total");
+
+        uint256 balanceBeforeTransfer = IERC20(_dealData.underlyingDealToken).balanceOf(address(this));
+        IERC20(_dealData.underlyingDealToken).transferFrom(msg.sender, address(this), _depositUnderlyingAmount);
+        uint256 balanceAfterTransfer = IERC20(_dealData.underlyingDealToken).balanceOf(address(this));
+        uint256 underlyingDealTokenAmount = balanceAfterTransfer - balanceBeforeTransfer;
+
+        if (balanceAfterTransfer >= _dealData.underlyingDealTokenTotal) {
+            underlyingDepositComplete = true;
+            purchaseExpiry = block.timestamp + _dealData.purchaseDuration;
+            emit DealFullyFunded(address(this), block.timestamp, purchaseExpiry);
```

Comment:
we should initialise the escrow contract.

Reply:
Is there a reason to wait till the deal is fully funded to initialize the escrow contract? Is this something that could be initialized in the in initialize function of this contract?

Reply:
Because sometimes they can just create an upfrontdeal and not deposit anything. So if we create it during initialization, it's just a waste of contact creation.